File: 动态规划/nthUglyNumbers/nthUglyNumbers.py
#丑数为质因子为2、3、5的数，求解第n个丑数
#初始思路为循环，用一个数记录丑数个数，直到丑数个数达到退出
#上述解法的时间复杂度较高
#换种思路解法，类似于动态规划，维护3个指针，分别指向2、3、5的基数，则下一个丑数必定为num[i] * basej 的最小值，赋值后将指针向后移动一位即可求解

class Solution:
    def nthUglyNumber(self, n: int) -> int:
        items = [0,1]
        p2, p3, p5 = 1,1,1
        for i in range(2,n+1):
            num2,num3,num5 = items[p2]*2, items[p3]*3, items[p5]*5
            limit = min([num2, num3, num5])
            items.append(limit)
            if limit == num2:
                p2+=1
            if limit == num3:
                p3+=1
            if limit == num5:
                p5+=1
        return items[n]

File: 动态规划/nthUglyNumbers/nthUglyNumbers.c
int nthUglyNumber(int n){
    int items[n+1];
    items[0] = 0;
    items[1] = 1;
    int p2=1,p3=1,p5=1;
    for(int i=2;i<n+1;i++){
        int num2 = items[p2] * 2;
        int num3 = items[p3] * 3;
        int num5 = items[p5] * 5;
        int min = num2 > num5 ?  num5 : num2;
        min = min > num3 ? num3 : min;
        items[i] = min;
        if(min == num2){
            p2++;
        }
        if(min == num3){
            p3++;
        }
        if(min == num5){
            p5++;
        }
    }
    return items[n];
}